In [2]:
import geopandas as gp
import pandas as pd
import os
import numpy as np
import re
from collections import Counter

# Alabama 2024 Primary Election Results

In [3]:
path = './raw-from-source/2024 Primary Precinct Results/'

files = [i for i in os.listdir(path) if i != ".DS_Store"]

In [4]:
county_list = []
for file in files: 
    print(file)
    temp = pd.read_excel(path + file)
    county_name = file.split("-")[-1][0:-4]
    if county_name == "StClair":
        county_name = "St. Clair"
    temp["Party"] = temp["Party"].str.strip()
    temp["Party"] = temp["Party"].fillna("")
    temp["pivot_col"] = temp["Contest Title"].str.strip()+"-:-"+temp["Candidate"].str.strip()
    temp["pivot_col"] = np.where(temp["Party"]=="",temp["pivot_col"],temp["pivot_col"]+"-:-"+temp["Party"].str.strip())

    temp.drop(["Contest Title", "Party", "Candidate"], axis = 1, inplace = True)

    
    rename_dict = {i:i+"-:-"+county_name for i in temp.columns if i != "pivot_col"}
    temp.rename(columns = rename_dict, inplace = True)
    temp_transpose = temp.set_index("pivot_col").T
    temp_transpose.reset_index(inplace = True, drop = False)
    ser = temp_transpose["index"].value_counts()
    if ser[ser>1].shape[0]>0:
        raise ValueError
    temp_transpose["County"] = county_name
    #print(temp_transpose.head(10))
    #break
#     print(temp_transpose.head())
    county_list.append(temp_transpose)

2024-Primary-Chilton.xls
2024-Primary-Greene.xls
2024-Primary-Marion.xls
2024-Primary-Colbert.xls
2024-Primary-Clay.xls
2024-Primary-Bullock.xls
2024-Primary-Montgomery.xls
2024-Primary-Henry.xls
2024-Primary-Cherokee.xls
2024-Primary-Crenshaw.xls
2024-Primary-Winston.xls
2024-Primary-Cleburne.xls
2024-Primary-Madison.xls
2024-Primary-DeKalb.xls
2024-Primary-Walker.xls
2024-Primary-Choctaw.xls
2024-Primary-Pike.xls
2024-Primary-Monroe.xls
2024-Primary-Randolph.xls
2024-Primary-Chambers.xls
2024-Primary-Conecuh.xls
2024-Primary-Fayette.xls
2024-Primary-Wilcox.xls
2024-Primary-StClair.xls
2024-Primary-Escambia.xls
2024-Primary-Butler.xls
2024-Primary-Covington.xls
2024-Primary-Blount.xls
2024-Primary-Coffee.xls
2024-Primary-Russell.xls
2024-Primary-Baldwin.xls
2024-Primary-Barbour.xls
2024-Primary-Sumter.xls
2024-Primary-Limestone.xls
2024-Primary-Lamar.xls
2024-Primary-Tuscaloosa.xls
2024-Primary-Tallapoosa.xls
2024-Primary-Autauga.xls
2024-Primary-Houston.xls
2024-Primary-Hale.xls
2024

In [5]:
comb = pd.concat(county_list)

In [6]:
comb = comb.fillna(0)

In [7]:
comb.head()

pivot_col,index,REGISTERED VOTERS - TOTAL-:-Registered Voters - Total,BALLOTS CAST - TOTAL-:-Ballots Cast - Total,BALLOTS CAST - DEMOCRAT-:-Ballots Cast - Alabama Democratic P-:-DEM,BALLOTS CAST - REPUBLICAN-:-Ballots Cast - Alabama Republican P-:-REP,BALLOTS CAST - NON-PARTISAN-:-Ballots Cast - Nonpartisan,BALLOTS CAST - BLANK-:-Ballots Cast - Blank,PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr-:-DEM,PRESIDENT OF THE UNITED STATES-:-Ryan L. Binkley-:-REP,PRESIDENT OF THE UNITED STATES-:-Dean Phillips-:-DEM,...,"MEMBER, LAUDERDALE COUNTY COMMISSION, DISTRICT NO. 2-:-Over Votes-:-REP","MEMBER, LAUDERDALE COUNTY COMMISSION, DISTRICT NO. 2-:-Under Votes-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Randy Burns-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Anita Haddock Olbon-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Over Votes-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Under Votes-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Valerie Cook-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Mike McMillan-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Over Votes-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Under Votes-:-REP"
0,ABSENTEE-:-Chilton,0.0,166.0,7.0,159.0,0.0,0.0,6.0,0.0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,CANE CREEK GAP OF MTN-:-Chilton,0.0,354.0,4.0,350.0,0.0,0.0,3.0,0.0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,CLANTON REC_-:-Chilton,0.0,956.0,32.0,924.0,0.0,0.0,28.0,2.0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,COLLINS CHAPEL FS-:-Chilton,0.0,191.0,4.0,187.0,0.0,0.0,1.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,COURTHOUSE CLANTON-:-Chilton,0.0,499.0,20.0,479.0,0.0,0.0,16.0,1.0,2.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [8]:
comb[comb["index"].str.contains("TILLMAN")]

pivot_col,index,REGISTERED VOTERS - TOTAL-:-Registered Voters - Total,BALLOTS CAST - TOTAL-:-Ballots Cast - Total,BALLOTS CAST - DEMOCRAT-:-Ballots Cast - Alabama Democratic P-:-DEM,BALLOTS CAST - REPUBLICAN-:-Ballots Cast - Alabama Republican P-:-REP,BALLOTS CAST - NON-PARTISAN-:-Ballots Cast - Nonpartisan,BALLOTS CAST - BLANK-:-Ballots Cast - Blank,PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr-:-DEM,PRESIDENT OF THE UNITED STATES-:-Ryan L. Binkley-:-REP,PRESIDENT OF THE UNITED STATES-:-Dean Phillips-:-DEM,...,"MEMBER, LAUDERDALE COUNTY COMMISSION, DISTRICT NO. 2-:-Over Votes-:-REP","MEMBER, LAUDERDALE COUNTY COMMISSION, DISTRICT NO. 2-:-Under Votes-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Randy Burns-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Anita Haddock Olbon-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Over Votes-:-REP","MEMBER, LAUDERDALE COUNTY BOARD OF EDUCATION, DIST NO. 3-:-Under Votes-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Valerie Cook-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Mike McMillan-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Over Votes-:-REP","MEMBER, BIBB COUNTY BOARD OF EDUCATION, DISTRICT NO. 2-:-Under Votes-:-REP"
51,STILLMAN COLLEGE-:-Tuscaloosa,2506.0,549.0,518.0,28.0,3.0,0.0,483.0,0.0,10.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
83,TILLMAN'S CORNER COMM CTR CONG 1-:-Mobile,0.0,440.0,80.0,359.0,1.0,0.0,73.0,2.0,3.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
84,TILLMAN'S CORNER COMM CTR CONG 2-:-Mobile,0.0,749.0,97.0,652.0,0.0,0.0,76.0,0.0,7.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [9]:
comb_list = [i for i in comb.columns if ("REPRESENTATIVE" in i or "PRESIDENT" in i or "DELEGATES" in i or "STATEWIDE" in i or "CHIEF JUSTICE" in i or "APPEALS JUDGE" in i) and ("Under Votes" not in i and "Over Votes" not in i)]

In [10]:
comb_list.sort()

In [11]:
comb_list

['CHIEF JUSTICE OF THE SUPREME COURT-:-Bryan Taylor-:-REP',
 'CHIEF JUSTICE OF THE SUPREME COURT-:-Sarah Stewart-:-REP',
 'COURT OF CIVIL APPEALS JUDGE, PLACE 2-:-Chad Hanson-:-REP',
 'COURT OF CIVIL APPEALS JUDGE, PLACE 2-:-Stephen Davis Parker-:-REP',
 'COURT OF CRIMINAL APPEALS JUDGE, PLACE 2-:-Rich Anderson-:-REP',
 'COURT OF CRIMINAL APPEALS JUDGE, PLACE 2-:-Thomas Govan-:-REP',
 'DEMOCRATIC DELEGATES 1ST CONGRESSIONAL DISTRICT-:-Jason Fisher-:-DEM',
 'DEMOCRATIC DELEGATES 1ST CONGRESSIONAL DISTRICT-:-Joseph L. Johnson-:-DEM',
 'DEMOCRATIC DELEGATES 1ST CONGRESSIONAL DISTRICT-:-Joseph R Biden Jr Delegates-:-DEM',
 'DEMOCRATIC DELEGATES 1ST CONGRESSIONAL DISTRICT-:-Mary Parker-:-DEM',
 'DEMOCRATIC DELEGATES 2ND CONGRESSIONAL DISTRICT-:-Barabara Drummond-:-DEM',
 'DEMOCRATIC DELEGATES 2ND CONGRESSIONAL DISTRICT-:-Barbara Drummond-:-DEM',
 'DEMOCRATIC DELEGATES 2ND CONGRESSIONAL DISTRICT-:-Jim Lawlor-:-DEM',
 'DEMOCRATIC DELEGATES 2ND CONGRESSIONAL DISTRICT-:-Joseph R Biden Jr Delega

In [12]:
remaining = [i for i in list(comb.columns) if i not in comb_list]

In [15]:
[i for i in remaining if "COUNTY" not in i and "Over Votes" not in i and "Under Votes" not in i]

['index',
 'REGISTERED VOTERS - TOTAL-:-Registered Voters - Total',
 'BALLOTS CAST - TOTAL-:-Ballots Cast - Total',
 'BALLOTS CAST - DEMOCRAT-:-Ballots Cast - Alabama Democratic P-:-DEM',
 'BALLOTS CAST - REPUBLICAN-:-Ballots Cast - Alabama Republican P-:-REP',
 'BALLOTS CAST - NON-PARTISAN-:-Ballots Cast - Nonpartisan',
 'BALLOTS CAST - BLANK-:-Ballots Cast - Blank',
 'STATE BOARD OF EDUCATION MEMBER DISTRICT 3-:-Ann C. Eubank-:-REP',
 'STATE BOARD OF EDUCATION MEMBER DISTRICT 3-:-Charlotte Meadows-:-REP',
 'STATE BOARD OF EDUCATION MEMBER DISTRICT 3-:-Kelly Mooney-:-REP',
 'STATE BOARD OF EDUCATION MEMBER DISTRICT 3-:-Melissa B. Snowden-:-REP',
 'CIRCUIT COURT JUDGE, 19TH JUDICIAL CIRCUIT, PLACE 4-:-Dee Dee Calhoon-:-REP',
 'CIRCUIT COURT JUDGE, 19TH JUDICIAL CIRCUIT, PLACE 4-:-Nicole Clark-:-REP',
 'CIRCUIT COURT JUDGE, 19TH JUDICIAL CIRCUIT, PLACE 4-:-Jacquelyn Tomlinson-:-REP',
 'County',
 'CIRCUIT COURT JUDGE, 17TH JUDICIAL CIRCUIT, PLACE 1-:-Gregory S. Griggers-:-DEM',
 'CIRCUIT

In [12]:
deleg_dict = {i:"PRESIDENT OF THE UNITED STATES-:-"+i.split("-:-")[1]+"-:-"+i.split("-:-")[2] for i in comb_list if "DEMOCRATIC DELEGATES" in i}

In [13]:
comb = comb[["index"]+comb_list]

In [14]:
comb.rename(columns = deleg_dict, inplace = True)

In [15]:
clean_dict2 = {'PRESIDENT OF THE UNITED STATES-:-Albert Turner Jr':'PRESIDENT OF THE UNITED STATES-:-Albert Turner Jr.',
'PRESIDENT OF THE UNITED STATES-:-Barabara Drummond-:-DEM':'PRESIDENT OF THE UNITED STATES-:-Barbara Drummond-:-DEM',
 'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr -:-DEM':'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr-:-DEM',
 'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr-:-DEM':'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr-:-DEM',
 'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden, Jr-:-DEM':'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr-:-DEM',
 'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden,Jr-:-DEM':'PRESIDENT OF THE UNITED STATES-:-Joseph R Biden Jr-:-DEM',
 'UNITED STATES REPRESENTATIVE, 2ND CONGRESSIONAL DISTRICT-:-Caroleene Dobson-:-REP':'UNITED STATES REPRESENTATIVE, 2ND CONGRESSIONAL DISTRICT-:-Caroleen Dobson-:-REP',
  'UNITED STATES REPRESENTATIVE, 2ND CONGRESSIONAL DISTRICT-:-Juandalynn "Le Le" Givan-:-DEM':'UNITED STATES REPRESENTATIVE, 2ND CONGRESSIONAL DISTRICT-:-Juandalynn "Lele" Givan-:-DEM',
 'PRESIDENT OF THE UNITED STATES-:-Albert Turner Jr-:-DEM': 'PRESIDENT OF THE UNITED STATES-:-Albert Turner Jr.-:-DEM',}

In [16]:
comb.columns = [i.replace(" Delegates","").replace("For ","") for i in list(comb.columns)]

In [17]:
comb.rename(columns = clean_dict2, inplace = True)

In [18]:
comb2 = comb.groupby(level = 0,axis = 1).sum()

In [19]:
comb2.columns

Index(['CHIEF JUSTICE OF THE SUPREME COURT-:-Bryan Taylor-:-REP',
       'CHIEF JUSTICE OF THE SUPREME COURT-:-Sarah Stewart-:-REP',
       'COURT OF CIVIL APPEALS JUDGE, PLACE 2-:-Chad Hanson-:-REP',
       'COURT OF CIVIL APPEALS JUDGE, PLACE 2-:-Stephen Davis Parker-:-REP',
       'COURT OF CRIMINAL APPEALS JUDGE, PLACE 2-:-Rich Anderson-:-REP',
       'COURT OF CRIMINAL APPEALS JUDGE, PLACE 2-:-Thomas Govan-:-REP',
       'PRESIDENT OF THE UNITED STATES-:-Adarris May-:-DEM',
       'PRESIDENT OF THE UNITED STATES-:-Albert Turner Jr.-:-DEM',
       'PRESIDENT OF THE UNITED STATES-:-April Hodges-:-DEM',
       'PRESIDENT OF THE UNITED STATES-:-Barbara Drummond-:-DEM',
       'PRESIDENT OF THE UNITED STATES-:-Benard Simelton-:-DEM',
       'PRESIDENT OF THE UNITED STATES-:-Chris Christie-:-REP',
       'PRESIDENT OF THE UNITED STATES-:-Christopher John England-:-DEM',
       'PRESIDENT OF THE UNITED STATES-:-Ciara Smith-:-DEM',
       'PRESIDENT OF THE UNITED STATES-:-Curtis Travis-:-

In [20]:
def get_race(race_string):
    race_string = race_string.title()
    race_string = race_string.replace("(Vote For 1)","")
    if "U.S. House" in race_string or 'Us House' in race_string or "United States Representative" in race_string:
        return "CON"
    elif "Public Service" in race_string:
        if "1" in race_string:
            return "PS1"
        elif "2" in race_string:
            return "PS2"
        else:
            return "PSC"
    elif "United States Senator" in race_string:
        return "USS"
    elif "State House" in race_string or "State Representative" in race_string:
        return "SL"
    elif "State Senator" in race_string:
        return "SU"
    elif "President" in race_string:
        return "PRE"
    elif "US Senate" in race_string or "Us Senate" in race_string:
        return "USS"

    elif "Attorney General" in race_string:
        return "ATG"
    elif "Auditor General" in race_string or "State Auditor" in race_string:
        return "AUD"
    elif "Treasurer" in race_string:
        return "TRE"
    elif "Superintendent" in race_string:
        return "SUP"
    elif "Secretary Of State" in race_string:
        return "SOS"
    elif "Lieutenant Governor" in race_string:
        return "LTG"
    elif "Governor" in race_string:
        return "GOV"
    elif "Commissioner Of Labor" in race_string:
        return "LAB"
    elif "Commissioner Of Agriculture" in race_string:
        return "AGR"
    elif "Commissioner Of Insurance" in race_string:
        return "INS"
    elif "Associate Justice Of The Supreme Court" in race_string:
        if "5" in race_string:
            return "AJ5"
        elif "6" in race_string:
            return "AJ6"
    elif "Constitution" in race_string:
        return "CNS"
    elif "Chief Justice" in race_string:
        return "CFJ"
    elif "Amendment" in race_string:
        if "One " in race_string:
            return "A01"
        elif "Two " in race_string:
            return "A02"
        elif "Three " in race_string:
            return "A03"
        elif "Four " in race_string:
            return "A04"
        elif "Five " in race_string:
            return "A05"
        elif "Six " in race_string:
            return "A06"
        elif "Seven " in race_string:
            return "A07"
        elif "Eight " in race_string:
            return "A08"
        elif "Nine " in race_string:
            return "A09"
        elif "Ten " in race_string:
            return "A10"
        else:
            print("No race for:", race_string)
            raise ValueError
    elif "Referendum" in race_string:
        if "1" in race_string or "A" in race_string:
            return "RFA"
        elif "2" in race_string or "B" in race_string:
            return "RFB"
        else:
            print("No race for:", race_string)
            raise ValueError
    elif "Civil Appeals Judge" in race_string:
        if "1" in race_string:
            return "CV1"
        elif "2" in race_string:
            return "CV2"
    elif "Criminal Appeals Judge" in race_string:
        if "1" in race_string:
            return "CR1"
        elif "2" in race_string:
            return "CR2"
    else:
        print("No race for:", race_string)
        raise ValueError
        
def get_election_type(race_string):
    if "(runoff)" in race_string:
        return "R"
    else:
        return "G"
        
def get_party(race_string):
    if "REP" in race_string:
        return "R"
    elif "DEM" in race_string or "(Democrat)" in race_string:
        return "D"
    elif "LIB" in race_string or "(L)" in race_string:
        return "L"
    elif "NON" in race_string:
        return "O"
    elif "IND" in race_string:
        return "I"
    elif race_string[0:3]=="Yes":
        return "YES"
    elif race_string[0:2]=="No":
        return "NO"
    else:
        print("NO RACE", race_string)
        return ""
           
def get_name(name_string):
    if "Write-In" in name_string:
        return "WRI"
    if "AMENDMENT" not in name_string and "Referendum" not in name_string and "CONSTITUTION" not in name_string:
        #print(name_string)
        #name_string = name_string.split(" (")[0]
        name_string = name_string.replace("'","")
        likely_last = name_string.split(" ")[-1]
        proposed_last = likely_last[:3]
        if proposed_last in ['II', 'III', 'Jr', 'Jr.', 'Sr.', 'JR.', "JR", "IV","Jr-"]:
            likely_last = name_string.split(" ")[-2]
            proposed_last = likely_last[:3]
        #print(proposed_last.upper())
        return proposed_last.upper()
    else:
        return name_string.split("-:-")[1].upper()
#     name_string = name_string.split("-:-")[1]
#     name_string = name_string.replace(" (I)","")
#     name_string = name_string.replace("'","")
#     likely_last = name_string.split(" ")[-1]
#     proposed_last = likely_last[:3]
#     if proposed_last in ['II', 'III', 'Jr', 'Jr.', 'Sr.', 'JR.', "JR", "IV"]:
#         likely_last = name_string.split(" ")[-2]
#         proposed_last = likely_last[:3]
#     return proposed_last.upper()

def get_district(race_string, fill_level):
    race_string = race_string.split("-:-")[0]
    race_string = race_string.replace(" (Vote For 1)","")
    if "UNITED STATES REPRESENTATIVE" in race_string:
        break_word = "REPRESENTATIVE, "
        temp = race_string.split(break_word)[1]
        temp = re.findall('\d*', temp)[0]
    elif "STATE REPRESENTATIVE" in race_string or "STATE SENATOR" in race_string:
        break_word = "DISTRICT "
        temp = race_string.split(break_word)[1]
    else:
        raise ValueError
    
    return temp.zfill(fill_level)

def column_rename_function(name_string):
    election_type = "P"
    year = "24"
    party = get_party(name_string.split("-:-")[-1])
    race = get_race(name_string)
    district = ""
    if race in ["CON", "SU"]:
        district = get_district(name_string, 2)
        year = ""
    elif race in ["SL"]:
        district = get_district(name_string, 3)
        year = ""
    
    name = get_name(name_string)
    if "CONSTITUTION" in name_string or "STATEWIDE AMENDMENT" in name_string:

        new_col_name = election_type + year + race + district +  name
    else:
        new_col_name = election_type + year + race + district + party + name
        print(election_type)
        print(year)
        print(race)
        print(district)
        print(name)
    if len(new_col_name) > 10:
        print(name_string)
        print(new_col_name)
    return new_col_name

# Make a dictionary that points to the new column names and checks for duplicates
race_columns = [i for i in list(comb2.columns) if i != 'index']

race_updates_dict = {}
race_updates_reversed = {}
clean_dups = {}
new_names = []
for val in race_columns:
    new_name = column_rename_function(val)
    race_updates_dict[val] = new_name
    if new_name not in new_names:
        new_names.append(new_name)
        race_updates_reversed[new_name] = val
    else:
        print("Duplicate", new_name)
        print(race_updates_reversed[new_name])
        print(val)
        clean_dups[val] = race_updates_reversed[new_name]

P
24
CFJ

TAY
P
24
CFJ

STE
P
24
CV2

HAN
P
24
CV2

PAR
P
24
CR2

AND
P
24
CR2

GOV
P
24
PRE

MAY
P
24
PRE

TUR
P
24
PRE

HOD
P
24
PRE

DRU
P
24
PRE

SIM
P
24
PRE

CHR
P
24
PRE

ENG
P
24
PRE

SMI
P
24
PRE

TRA
P
24
PRE

SIL
P
24
PRE

STU
P
24
PRE

PHI
P
24
PRE

TRU
P
24
PRE

BEL
P
24
PRE

BAT
P
24
PRE

FIS
P
24
PRE

GRA
P
24
PRE

LAW
P
24
PRE

SQU
P
24
PRE

JOH
P
24
PRE

BID
P
24
PRE

JUA
P
24
PRE

WIL
P
24
PRE

LAW
Duplicate P24PREDLAW
PRESIDENT OF THE UNITED STATES-:-Jim Lawlor-:-DEM
PRESIDENT OF THE UNITED STATES-:-Kelvin Lawrence-:-DEM
P
24
PRE

LEW
P
24
PRE

HAT
P
24
PRE

ADA
P
24
PRE

PIE
P
24
PRE

HAL
P
24
PRE

COL
P
24
PRE

JON
P
24
PRE

PAR
P
24
PRE

HAL
P
24
PRE

HOW
P
24
PRE

MCC
P
24
PRE

EDI
P
24
PRE

ENS
P
24
PRE

HAD
P
24
PRE

STE
P
24
PRE

LIN
P
24
PRE

HOL
P
24
PRE

DES
P
24
PRE

BIN
P
24
PRE

TYS
P
24
PRE

MCK
P
24
PRE

STA
P
24
PRE

STA
P
24
PRE

RAM
P
24
PRE

FIG
P
24
PRE

YOU
P
24
PRE

KRA
P
24
PSC

MCC
P
24
PSC

CAV
P

CON
01
MOO
P

CON
01
CAR
P

CON
02
DAN
P

CON

In [21]:
race_updates_dict["PRESIDENT OF THE UNITED STATES-:-Jim Lawlor-:-DEM"] = "P24PREDJLA"
race_updates_dict['PRESIDENT OF THE UNITED STATES-:-Uncommitted-:-DEM'] = 'P24PREDUNC'
race_updates_dict['PRESIDENT OF THE UNITED STATES-:-Uncommitted-:-REP'] = 'P24PRERUNC'

In [22]:
myKeys = list(race_updates_dict.values())
myKeys.sort()
sorted_dict = dict(sorted(race_updates_dict.items(), key=lambda x:x[1]))
export_dict = {i:key for key, i in sorted_dict.items()}

In [23]:
pd.DataFrame(export_dict.items()).to_csv("./field_names_prim.csv", index = False)


In [24]:
comb = comb2.copy(deep = True)

In [25]:
comb

,CHIEF JUSTICE OF THE SUPREME COURT-:-Bryan Taylor-:-REP,CHIEF JUSTICE OF THE SUPREME COURT-:-Sarah Stewart-:-REP,"COURT OF CIVIL APPEALS JUDGE, PLACE 2-:-Chad Hanson-:-REP","COURT OF CIVIL APPEALS JUDGE, PLACE 2-:-Stephen Davis Parker-:-REP","COURT OF CRIMINAL APPEALS JUDGE, PLACE 2-:-Rich Anderson-:-REP","COURT OF CRIMINAL APPEALS JUDGE, PLACE 2-:-Thomas Govan-:-REP",PRESIDENT OF THE UNITED STATES-:-Adarris May-:-DEM,PRESIDENT OF THE UNITED STATES-:-Albert Turner Jr.-:-DEM,PRESIDENT OF THE UNITED STATES-:-April Hodges-:-DEM,PRESIDENT OF THE UNITED STATES-:-Barbara Drummond-:-DEM,...,"UNITED STATES REPRESENTATIVE, 4TH CONGRESSIONAL DISTRICT-:-Justin Holcomb-:-REP","UNITED STATES REPRESENTATIVE, 4TH CONGRESSIONAL DISTRICT-:-Robert B. Aderholt-:-REP","UNITED STATES REPRESENTATIVE, 6TH CONGRESSIONAL DISTRICT-:-Gary Palmer-:-REP","UNITED STATES REPRESENTATIVE, 6TH CONGRESSIONAL DISTRICT-:-Gerrick Wilkins-:-REP","UNITED STATES REPRESENTATIVE, 6TH CONGRESSIONAL DISTRICT-:-Ken McFeeters-:-REP","UNITED STATES REPRESENTATIVE, 7TH CONGRESSIONAL DISTRICT-:-Chris Davis-:-DEM","UNITED STATES REPRESENTATIVE, 7TH CONGRESSIONAL DISTRICT-:-Christian Horn-:-REP","UNITED STATES REPRESENTATIVE, 7TH CONGRESSIONAL DISTRICT-:-Robin Litaker-:-REP","UNITED STATES REPRESENTATIVE, 7TH CONGRESSIONAL DISTRICT-:-Terri A. Sewell-:-DEM",index
0,37.0,89.0,45.0,40.0,46.0,38.0,5.0,0.0,0.0,0.0,...,0.0,0.0,118.0,19.0,2.0,0.0,0.0,0.0,0.0,ABSENTEE-:-Chilton
1,95.0,211.0,150.0,114.0,128.0,138.0,2.0,0.0,0.0,0.0,...,0.0,0.0,277.0,27.0,13.0,0.0,0.0,0.0,0.0,CANE CREEK GAP OF MTN-:-Chilton
2,291.0,556.0,424.0,325.0,339.0,422.0,23.0,0.0,0.0,0.0,...,0.0,0.0,711.0,148.0,22.0,0.0,0.0,0.0,0.0,CLANTON REC_-:-Chilton
3,60.0,98.0,75.0,62.0,53.0,81.0,1.0,0.0,0.0,0.0,...,0.0,0.0,149.0,25.0,5.0,0.0,0.0,0.0,0.0,COLLINS CHAPEL FS-:-Chilton
4,151.0,274.0,208.0,166.0,165.0,219.0,12.0,0.0,0.0,0.0,...,0.0,0.0,378.0,55.0,16.0,0.0,0.0,0.0,0.0,COURTHOUSE CLANTON-:-Chilton
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5,199.0,333.0,254.0,226.0,255.0,232.0,12.0,0.0,0.0,0.0,...,0.0,0.0,489.0,32.0,31.0,0.0,0.0,0.0,0.0,GREENPOND FIRE DEPT-:-Bibb
6,40.0,117.0,76.0,62.0,58.0,87.0,12.0,0.0,0.0,0.0,...,0.0,0.0,145.0,16.0,4.0,0.0,0.0,0.0,0.0,LAWLEY COMM CTR-:-Bibb
7,162.0,333.0,250.0,188.0,170.0,268.0,27.0,0.0,0.0,0.0,...,0.0,0.0,436.0,54.0,28.0,0.0,0.0,0.0,0.0,NATIONAL GUARD-:-Bibb
8,3.0,0.0,1.0,2.0,3.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,2.0,0.0,1.0,0.0,0.0,0.0,0.0,PROVISIONAL-:-Bibb


In [26]:
myKeys = list(race_updates_dict.values())
myKeys.sort()
#sorted_dict = {myKeys: race_updates_dict[i] for i in myKeys}


sorted_dict = dict(sorted(race_updates_dict.items(), key=lambda x:x[1]))
comb.rename(columns = race_updates_dict, inplace = True)


In [27]:
comb

,P24CFJRTAY,P24CFJRSTE,P24CV2RHAN,P24CV2RPAR,P24CR2RAND,P24CR2RGOV,P24PREDMAY,P24PREDTUR,P24PREDHOD,P24PREDDRU,...,PCON04RHOL,PCON04RADE,PCON06RPAL,PCON06RWIL,PCON06RMCF,PCON07DDAV,PCON07RHOR,PCON07RLIT,PCON07DSEW,index
0,37.0,89.0,45.0,40.0,46.0,38.0,5.0,0.0,0.0,0.0,...,0.0,0.0,118.0,19.0,2.0,0.0,0.0,0.0,0.0,ABSENTEE-:-Chilton
1,95.0,211.0,150.0,114.0,128.0,138.0,2.0,0.0,0.0,0.0,...,0.0,0.0,277.0,27.0,13.0,0.0,0.0,0.0,0.0,CANE CREEK GAP OF MTN-:-Chilton
2,291.0,556.0,424.0,325.0,339.0,422.0,23.0,0.0,0.0,0.0,...,0.0,0.0,711.0,148.0,22.0,0.0,0.0,0.0,0.0,CLANTON REC_-:-Chilton
3,60.0,98.0,75.0,62.0,53.0,81.0,1.0,0.0,0.0,0.0,...,0.0,0.0,149.0,25.0,5.0,0.0,0.0,0.0,0.0,COLLINS CHAPEL FS-:-Chilton
4,151.0,274.0,208.0,166.0,165.0,219.0,12.0,0.0,0.0,0.0,...,0.0,0.0,378.0,55.0,16.0,0.0,0.0,0.0,0.0,COURTHOUSE CLANTON-:-Chilton
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5,199.0,333.0,254.0,226.0,255.0,232.0,12.0,0.0,0.0,0.0,...,0.0,0.0,489.0,32.0,31.0,0.0,0.0,0.0,0.0,GREENPOND FIRE DEPT-:-Bibb
6,40.0,117.0,76.0,62.0,58.0,87.0,12.0,0.0,0.0,0.0,...,0.0,0.0,145.0,16.0,4.0,0.0,0.0,0.0,0.0,LAWLEY COMM CTR-:-Bibb
7,162.0,333.0,250.0,188.0,170.0,268.0,27.0,0.0,0.0,0.0,...,0.0,0.0,436.0,54.0,28.0,0.0,0.0,0.0,0.0,NATIONAL GUARD-:-Bibb
8,3.0,0.0,1.0,2.0,3.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,2.0,0.0,1.0,0.0,0.0,0.0,0.0,PROVISIONAL-:-Bibb


In [28]:
comb = comb[["index"] + myKeys]

In [29]:
comb

,index,P24A01NO,P24A01YES,P24CFJRSTE,P24CFJRTAY,P24CR2RAND,P24CR2RGOV,P24CV2RHAN,P24CV2RPAR,P24PREDADA,...,PCON03RROG,PCON04RADE,PCON04RHOL,PCON06RMCF,PCON06RPAL,PCON06RWIL,PCON07DDAV,PCON07DSEW,PCON07RHOR,PCON07RLIT
0,ABSENTEE-:-Chilton,77.0,48.0,89.0,37.0,46.0,38.0,45.0,40.0,0.0,...,0.0,0.0,0.0,2.0,118.0,19.0,0.0,0.0,0.0,0.0
1,CANE CREEK GAP OF MTN-:-Chilton,159.0,145.0,211.0,95.0,128.0,138.0,150.0,114.0,0.0,...,0.0,0.0,0.0,13.0,277.0,27.0,0.0,0.0,0.0,0.0
2,CLANTON REC_-:-Chilton,417.0,417.0,556.0,291.0,339.0,422.0,424.0,325.0,0.0,...,0.0,0.0,0.0,22.0,711.0,148.0,0.0,0.0,0.0,0.0
3,COLLINS CHAPEL FS-:-Chilton,91.0,77.0,98.0,60.0,53.0,81.0,75.0,62.0,0.0,...,0.0,0.0,0.0,5.0,149.0,25.0,0.0,0.0,0.0,0.0
4,COURTHOUSE CLANTON-:-Chilton,220.0,210.0,274.0,151.0,165.0,219.0,208.0,166.0,0.0,...,0.0,0.0,0.0,16.0,378.0,55.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5,GREENPOND FIRE DEPT-:-Bibb,313.0,224.0,333.0,199.0,255.0,232.0,254.0,226.0,0.0,...,0.0,0.0,0.0,31.0,489.0,32.0,0.0,0.0,0.0,0.0
6,LAWLEY COMM CTR-:-Bibb,90.0,81.0,117.0,40.0,58.0,87.0,76.0,62.0,0.0,...,0.0,0.0,0.0,4.0,145.0,16.0,0.0,0.0,0.0,0.0
7,NATIONAL GUARD-:-Bibb,312.0,218.0,333.0,162.0,170.0,268.0,250.0,188.0,0.0,...,0.0,0.0,0.0,28.0,436.0,54.0,0.0,0.0,0.0,0.0
8,PROVISIONAL-:-Bibb,1.0,3.0,0.0,3.0,3.0,0.0,1.0,2.0,0.0,...,0.0,0.0,0.0,1.0,2.0,0.0,0.0,0.0,0.0,0.0


In [30]:
export_dict = {i:key for key, i in sorted_dict.items()}

In [31]:
pd.DataFrame(export_dict.items()).to_csv("./field_names.csv", index = False)

In [32]:
comb["County"] = comb["index"].apply(lambda x:x.split("-:-")[1])

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/3688505148.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb["County"] = comb["index"].apply(lambda x:x.split("-:-")[1])


In [33]:
comb.groupby("County").sum().to_csv("./al_county_totals.csv")

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/1310031185.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  comb.groupby("County").sum().to_csv("./al_county_totals.csv")


In [34]:
# for i in list(sorted_dict.values()):
#     comb[i] = comb[i].astype(int)
#     print(i, sum(comb[i]))
    

P24A01NO 360944
P24A01YES 342511
P24CFJRSTE 335114
P24CFJRTAY 209208
P24CR2RAND 259919
P24CR2RGOV 210503
P24CV2RHAN 264894
P24CV2RPAR 205086
P24PREDADA 18482
P24PREDBAT 11230
P24PREDBEL 25449
P24PREDBID 168012
P24PREDCOL 31531
P24PREDDRU 24902
P24PREDEDI 18981
P24PREDENG 26736
P24PREDENS 19844
P24PREDFIG 36959
P24PREDFIS 6417
P24PREDGRA 8255
P24PREDHAD 8936
P24PREDHAL 12764
P24PREDHAT 20311
P24PREDHOD 9655
P24PREDHOL 26857
P24PREDHOW 6998
P24PREDJLA 1086
P24PREDJOH 6486
P24PREDJON 10902
P24PREDJUA 1299
P24PREDKRA 671
P24PREDLAW 24514
P24PREDLEW 10559
P24PREDLIN 765
P24PREDMAY 10913
P24PREDMCC 23273
P24PREDMCK 5095
P24PREDPAR 7419
P24PREDPHI 8440
P24PREDPIE 6309
P24PREDSIL 9960
P24PREDSIM 9113
P24PREDSMI 7348
P24PREDSQU 333
P24PREDSTA 11279
P24PREDSTE 26583
P24PREDTRA 4409
P24PREDTUR 28900
P24PREDTYS 36609
P24PREDWIL 1560
P24PREDYOU 22481
P24PRERBIN 509
P24PRERCHR 1442
P24PRERDES 8451
P24PRERHAL 77983
P24PRERRAM 1864
P24PRERSTA 9806
P24PRERSTU 752
P24PRERTRU 499121
P24PSCRCAV 330472
P24PSCRMCC 210126
PCON01RCAR 50312
PCON01RMOO 53956
PCON02DAVE 1623
PCON02DBRA 9010
PCON02DCOL 3445
PCON02DDAN 12879
PCON02DFIG 24979
PCON02DGIV 1260
PCON02DGRA 1580
PCON02DHAR 2007
PCON02DLEN 199
PCON02DPAT 289
PCON02DSIM 247
PCON02RALB 14470
PCON02RBRE 22589
PCON02RDOB 15102
PCON02RDUP 823
PCON02RGIL 838
PCON02RHAR 1415
PCON02RSHE 773
PCON02RTHO 1082
PCON03RBEV 5056
PCON03RNEW 10926
PCON03RROG 71242
PCON04RADE 88976
PCON04RHOL 22681
PCON06RMCF 5704
PCON06RPAL 76469
PCON06RWIL 9700
PCON07DDAV 4712
PCON07DSEW 59091
PCON07RHOR 18111
PCON07RLIT 12986

In [35]:
county = pd.read_csv("/Users/peterhorton/Documents/RDH/raw_data/census/PL_COUNTYFP_NAMES.csv")
al_cnty_fips_dict = dict((zip(county[county["STUSAB"]=="AL"]["NAME"],county[county["STUSAB"]=="AL"]["COUNTYFP"].astype(str).str.zfill(3))))

In [36]:
comb["Precinct"] = comb["index"].apply(lambda x: x.split("-:-")[0])

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/545608281.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb["Precinct"] = comb["index"].apply(lambda x: x.split("-:-")[0])


In [37]:
comb["County"] = comb["index"].apply(lambda x: x.split("-:-")[1])

comb["County_map"] = comb["County"] + " County"

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/2491721802.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb["County"] = comb["index"].apply(lambda x: x.split("-:-")[1])
/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/2491721802.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb["County_map"] = comb["County"] + " County"


In [38]:
comb["COUNTYFP"] = comb["County_map"].map(al_cnty_fips_dict).fillna(comb["County"])

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/3302747155.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb["COUNTYFP"] = comb["County_map"].map(al_cnty_fips_dict).fillna(comb["County"])


In [39]:
comb["index"] = comb["index"].apply(lambda x: x.split("-:-")[1]+"-:-"+x.split("-:-")[0])

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/3319974299.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb["index"] = comb["index"].apply(lambda x: x.split("-:-")[1]+"-:-"+x.split("-:-")[0])


In [40]:
comb.rename(columns = {"index":"UNIQUE_ID"}, inplace = True)

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/3538977619.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb.rename(columns = {"index":"UNIQUE_ID"}, inplace = True)


In [41]:
comb = comb[["UNIQUE_ID","COUNTYFP","County","Precinct"]+list(sorted_dict.values())]

In [42]:
comb["COUNTYFP"].unique()

array(['021', '063', '093', '033', '027', '011', '101', '067', '019',
       '041', '133', '029', '089', '049', '127', '023', '109', '099',
       '111', '017', '035', '057', '131', '115', '053', '013', '039',
       '009', '031', '113', '003', '005', '119', '083', '075', '125',
       '123', '001', '069', '065', '045', '095', '107', '081', '073',
       '025', '047', '129', '085', '087', '059', '105', '061', '043',
       '103', '037', '051', '117', '055', '097', '091', '079', '015',
       '121', '071', '077', '007'], dtype=object)

In [43]:
al_cnty_fips_dict

{'Autauga County': '001',
 'Baldwin County': '003',
 'Barbour County': '005',
 'Bibb County': '007',
 'Blount County': '009',
 'Bullock County': '011',
 'Butler County': '013',
 'Calhoun County': '015',
 'Chambers County': '017',
 'Cherokee County': '019',
 'Chilton County': '021',
 'Choctaw County': '023',
 'Clarke County': '025',
 'Clay County': '027',
 'Cleburne County': '029',
 'Coffee County': '031',
 'Colbert County': '033',
 'Conecuh County': '035',
 'Coosa County': '037',
 'Covington County': '039',
 'Crenshaw County': '041',
 'Cullman County': '043',
 'Dale County': '045',
 'Dallas County': '047',
 'DeKalb County': '049',
 'Elmore County': '051',
 'Escambia County': '053',
 'Etowah County': '055',
 'Fayette County': '057',
 'Franklin County': '059',
 'Geneva County': '061',
 'Greene County': '063',
 'Hale County': '065',
 'Henry County': '067',
 'Houston County': '069',
 'Jackson County': '071',
 'Jefferson County': '073',
 'Lamar County': '075',
 'Lauderdale County': '077',
 

In [44]:
comb["COUNTYFP"].unique()

array(['021', '063', '093', '033', '027', '011', '101', '067', '019',
       '041', '133', '029', '089', '049', '127', '023', '109', '099',
       '111', '017', '035', '057', '131', '115', '053', '013', '039',
       '009', '031', '113', '003', '005', '119', '083', '075', '125',
       '123', '001', '069', '065', '045', '095', '107', '081', '073',
       '025', '047', '129', '085', '087', '059', '105', '061', '043',
       '103', '037', '051', '117', '055', '097', '091', '079', '015',
       '121', '071', '077', '007'], dtype=object)

In [45]:
comb["UNIQUE_ID"] = comb["COUNTYFP"] + "-:-" + comb["Precinct"]

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/303240413.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb["UNIQUE_ID"] = comb["COUNTYFP"] + "-:-" + comb["Precinct"]


In [46]:
comb.sort_values(["COUNTYFP","Precinct"], inplace = True)

/var/folders/1t/0q4w6hm92mg_zxd84dfxmq3m0000gn/T/ipykernel_36961/2103180171.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  comb.sort_values(["COUNTYFP","Precinct"], inplace = True)


In [47]:
xl = pd.ExcelFile("./raw-from-source/totals_for_checks/RepublicanPartyOfficial-2024PrimaryResult-Amended (1).xlsx")

counties = [i for i in xl.sheet_names if i!="Summary"]

## Check County Totals

In [48]:
li_li = []

for county in counties:
        
    l = pd.read_excel("./raw-from-source/totals_for_checks/RepublicanPartyOfficial-2024PrimaryResult-Amended (1).xlsx", sheet_name = county)
    l.drop(l.columns[list(range(3,l.shape[1]))], axis = 1, inplace = True)

    contest = ""


    for key, val in l.iterrows():
        if val[1] == "Votes":
            contest = val[0]
        elif val[0] != "Total" and not pd.isna(val[0]):
            li_li.append([county, contest, val[0], val[1]])


In [49]:
county_totals = pd.DataFrame(li_li)
county_totals["id"] = county_totals[1]+"-"+county_totals[2]
county_totals = pd.pivot_table(county_totals, index = 0, columns = ["id"], values = 3)
county_totals.reset_index(inplace = True, drop = False)
county_totals = county_totals.fillna(0)
county_totals.to_csv("./al_rep_county_totals.csv", index = False)

In [50]:
county_totals

id,0,Alabama Chief Justice-Bryan Taylor,Alabama Chief Justice-Sarah Stewart,"Court of Civil Appeals Judge, Place 2-Chad Hanson","Court of Civil Appeals Judge, Place 2-Stephen Davis Parker","Court of Criminal Appeals Judge, Place 2-Rich Anderson","Court of Criminal Appeals Judge, Place 2-Thomas Govan",President of the Public Service Commission-Robert L. McCollum,President of the Public Service Commission-Twinkle Andress Cavanaugh,President-Chris Christie,...,US House 3rd District-Barron Rae Bevels,US House 3rd District-Bryan K. Newell,US House 3rd District-Mike Rogers,US House 4 District-Justin Holcomb,US House 4 District-Robert B. Aderholt,US House 6th District-Gary Palmer,US House 6th District-Gerrick Wilkins,US House 6th District-Ken McFeeters,US House 7th District-Christian Horn,US House 7th District-Robin Litaker
0,Autauga,3819.0,4660.0,4005.0,3157.0,4008.0,3216.0,2974.0,5593.0,17.0,...,0.0,0.0,0.0,0.0,0.0,6315.0,1157.0,651.0,0.0,0.0
1,Baldwin,12890.0,21455.0,18446.0,11357.0,17391.0,13148.0,13643.0,19895.0,79.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,Barbour,872.0,1065.0,939.0,777.0,953.0,746.0,750.0,1189.0,5.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,Bibb,939.0,1879.0,1369.0,1148.0,1160.0,1369.0,1087.0,1754.0,2.0,...,0.0,0.0,0.0,0.0,0.0,2520.0,257.0,133.0,0.0,0.0
4,Blount,3140.0,5790.0,4308.0,2608.0,4330.0,3614.0,4373.0,4635.0,17.0,...,0.0,0.0,0.0,2142.0,7167.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
62,Tuscaloosa,5832.0,10746.0,8153.0,6366.0,7562.0,7099.0,6422.0,10159.0,50.0,...,0.0,0.0,0.0,2146.0,7501.0,0.0,0.0,0.0,3910.0,2440.0
63,Walker,3536.0,7654.0,4689.0,5104.0,5391.0,4501.0,4886.0,6314.0,26.0,...,0.0,0.0,0.0,2169.0,9580.0,0.0,0.0,0.0,0.0,0.0
64,Washington,1071.0,1732.0,1447.0,977.0,1667.0,751.0,1069.0,1640.0,9.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
65,Wilcox,217.0,346.0,260.0,244.0,274.0,229.0,195.0,368.0,45.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,290.0,203.0


In [51]:
comb

,UNIQUE_ID,COUNTYFP,County,Precinct,P24A01NO,P24A01YES,P24CFJRSTE,P24CFJRTAY,P24CR2RAND,P24CR2RGOV,...,PCON03RROG,PCON04RADE,PCON04RHOL,PCON06RMCF,PCON06RPAL,PCON06RWIL,PCON07DDAV,PCON07DSEW,PCON07RHOR,PCON07RLIT
0,001-:-10 JONES COMM_ CTR_,001,Autauga,10 JONES COMM_ CTR_,69.0,72.0,62.0,54.0,63.0,44.0,...,0.0,0.0,0.0,14.0,84.0,20.0,0.0,0.0,0.0,0.0
1,001-:-100 TRINITY METHODIST,001,Autauga,100 TRINITY METHODIST,437.0,383.0,439.0,379.0,357.0,302.0,...,0.0,0.0,0.0,56.0,604.0,120.0,0.0,0.0,0.0,0.0
2,001-:-110 CENTRAL AL ELECTRIC,001,Autauga,110 CENTRAL AL ELECTRIC,46.0,70.0,39.0,20.0,24.0,26.0,...,0.0,0.0,0.0,3.0,51.0,4.0,0.0,0.0,0.0,0.0
3,001-:-140 AUTAUGAVILLE VFD,001,Autauga,140 AUTAUGAVILLE VFD,133.0,131.0,122.0,77.0,85.0,87.0,...,0.0,0.0,0.0,6.0,164.0,19.0,0.0,0.0,0.0,0.0
4,001-:-150 PRATTMONT BAPTIST,001,Autauga,150 PRATTMONT BAPTIST,140.0,156.0,152.0,114.0,134.0,101.0,...,0.0,0.0,0.0,21.0,180.0,63.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15,133-:-NESMITH CH,133,Winston,NESMITH CH,78.0,34.0,80.0,39.0,56.0,43.0,...,0.0,105.0,23.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
16,133-:-OLD UNION CHURCH,133,Winston,OLD UNION CHURCH,11.0,28.0,30.0,7.0,12.0,16.0,...,0.0,34.0,6.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
17,133-:-PEBBLE COMM_ CTR_,133,Winston,PEBBLE COMM_ CTR_,54.0,36.0,67.0,48.0,52.0,40.0,...,0.0,118.0,12.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
18,133-:-PROVISIONAL,133,Winston,PROVISIONAL,4.0,4.0,4.0,5.0,5.0,3.0,...,0.0,8.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [52]:
comb.to_csv("./al_prim_2024_prec_csv.csv", index = False)

In [53]:
comb

,UNIQUE_ID,COUNTYFP,County,Precinct,P24A01NO,P24A01YES,P24CFJRSTE,P24CFJRTAY,P24CR2RAND,P24CR2RGOV,...,PCON03RROG,PCON04RADE,PCON04RHOL,PCON06RMCF,PCON06RPAL,PCON06RWIL,PCON07DDAV,PCON07DSEW,PCON07RHOR,PCON07RLIT
0,001-:-10 JONES COMM_ CTR_,001,Autauga,10 JONES COMM_ CTR_,69.0,72.0,62.0,54.0,63.0,44.0,...,0.0,0.0,0.0,14.0,84.0,20.0,0.0,0.0,0.0,0.0
1,001-:-100 TRINITY METHODIST,001,Autauga,100 TRINITY METHODIST,437.0,383.0,439.0,379.0,357.0,302.0,...,0.0,0.0,0.0,56.0,604.0,120.0,0.0,0.0,0.0,0.0
2,001-:-110 CENTRAL AL ELECTRIC,001,Autauga,110 CENTRAL AL ELECTRIC,46.0,70.0,39.0,20.0,24.0,26.0,...,0.0,0.0,0.0,3.0,51.0,4.0,0.0,0.0,0.0,0.0
3,001-:-140 AUTAUGAVILLE VFD,001,Autauga,140 AUTAUGAVILLE VFD,133.0,131.0,122.0,77.0,85.0,87.0,...,0.0,0.0,0.0,6.0,164.0,19.0,0.0,0.0,0.0,0.0
4,001-:-150 PRATTMONT BAPTIST,001,Autauga,150 PRATTMONT BAPTIST,140.0,156.0,152.0,114.0,134.0,101.0,...,0.0,0.0,0.0,21.0,180.0,63.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15,133-:-NESMITH CH,133,Winston,NESMITH CH,78.0,34.0,80.0,39.0,56.0,43.0,...,0.0,105.0,23.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
16,133-:-OLD UNION CHURCH,133,Winston,OLD UNION CHURCH,11.0,28.0,30.0,7.0,12.0,16.0,...,0.0,34.0,6.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
17,133-:-PEBBLE COMM_ CTR_,133,Winston,PEBBLE COMM_ CTR_,54.0,36.0,67.0,48.0,52.0,40.0,...,0.0,118.0,12.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
18,133-:-PROVISIONAL,133,Winston,PROVISIONAL,4.0,4.0,4.0,5.0,5.0,3.0,...,0.0,8.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
